server/lib/pbench/indexer.py

```diff
@@ -1040,27 +1057,380 @@ def get_result_json_dirs(ptb):
+                   ...
+                },
+                "sample": {
+                    "@idx": 0,
+                    "measurement_idx": 0,
+                    "measurement_title": "Tpm",
+                    "measurement_type": "number",
+                    "name": "sample1"
+                }
+            }
+        """
+        name = self.user_benchmark["name"]
+        version = self.user_benchmark["version"]
+        driver = self.user_benchmark["driver"]
+        start_ts_col = driver["start_ts_col"]
+        end_ts_col = driver["end_ts_col"]
+        val_col = driver["val_col"]
+        methods = driver["methods"]
+        columns = driver["columns"]
+        columns_set = set(columns.keys())
+        bm_md_cols = driver["benchmark_md_cols"]
+        sam_md_cols = driver["sample_md_cols"]
+        ts_format = driver["ts_format"]
+        for iter_obj in self.ptb.get_iterations():
+            iteration_md = _dict_const(
+                [("name", iter_obj.name), ("number", iter_obj.number)]
+            )
+            for sample_obj in self.ptb.get_samples(iter_obj):
+                csv_name = os.path.join(sample_obj.path, "user-benchmark-result.csv")
+                try:
+                    with open(csv_name, "r") as fp:
+                        csv_reader = csv.DictReader(fp)
+                        row = next(csv_reader)
+                        # Now that we have read the first line of the .csv
+                        # file, the CSV reader now knows all the field names,
+                        # so we can compare against the expected field names.
+                        if set(csv_reader.fieldnames) - columns_set:
+                            self.ptb.idxctx.logger.warning(
+                                "Unexpected column(s) found, '{!r}' in {}",
+                                set(csv_reader.fieldnames) - columns_set,
+                                csv_name,
+                            )
+                            continue
+                except OSError as exc:
+                    if exc.errno == errno.ENOENT:
+                        continue
+                except Exception as exc:
+                    self.ptb.idxctx.logger.warning(
+                        "Unable to read '{}', '{}'", csv_name, exc
+                    )
+                    continue
+                # First construct the base sample metadata common to sample
+                # documents and result documents.
+                base_sample_md = _dict_const(
+                    [
+                        ("@idx", 0),
+                        ("measurement_idx", 0),
+                        ("measurement_title", val_col),
+                        ("measurement_type", "number"),
+                        ("name", sample_obj.name),
+                    ]
+                )
+                sample_md = dict()
+                sample_md.update(base_sample_md)
+                sample_md["closest_sample"] = sample_obj.number
```

Comment:
Is this because we assume that there is only one sample? Otherwise, how do we know that this is the closest sample? Or is there some comparison somewhere that I missed?

Reply:
Yes, right now, `pbench-user-benchmark` only generates one sample per iteration.